### PR TITLE
Use Fused Kernels for Depthwise Striding Subsampler: up to 1.5x faster training

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

This PR implements triton kernels for the subsampler in `ConformerEncoder`, profiling showed that the subsampler is heavily bottlenecked by memory bandwidth, so the fused kernels focus on reducing memory bandwidth by removing intermediary tensors

The first kernel implements:
```
mel input: [B, T, F, 1]
    ⇩
conv0 output: [B, T/2, F/2, 256] 
    ⇩
ReLU: [B, T/2, F/2, 256] 
    ⇩
Masking: [B, T/2, F/2, 256] 
    ⇩
conv_dw1 output: [B, T/4, F/4, 256]
```
it cuts memory bandwidth by more than 350x and even more in causal mode since it includes an extra padding layer, this fusion is not possible with `torch.compile` since it does not fuse the activation, the bias, or the masking into the convolution kernels, this kernel is most of the speedup

The second kernel is a general depthwise `conv2d` kernel, it's faster than torch kernel in forward since it fuses the pre and post masking and the causal padding, it has the most impact on M and L models, almost no impact on XL models because the encoder layers dominate the runtime, it is also faster than torch backward by 2x for the same reason

below are the speedup numbers expressed in relative speedup in E2E encoder step fwd+bwd, it uses the optimization in #16052 in both arms of the comparison

**Eager**

| size | fwd   | bwd   | e2e       |
| ---- | ----- | ----- | --------- |
| M    | 1.38x | 1.44x | **1.42x** |
| L    | 1.36x | 1.36x | **1.36x** |
| XL   | 1.05x | 1.20x | **1.14x** |

**torch.compile (max-autotune-no-cudagraphs)**

| size | fwd   | bwd   | e2e       |
| ---- | ----- | ----- | --------- |
| M    | 1.28x | 1.63x | **1.50x** |
| L    | 1.22x | 1.47x | **1.39x** |
| XL   | 1.11x | 1.23x | **1.19x** |

**Collection**: ASR

# Usage

Defaults to enabled if triton is installed, otherwise disabled, can also be controlled using the training config:
```yaml
model:
  encoder:
    use_triton: false
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

Trusted PRs run automatically through copy-pr-bot. For an untrusted PR, a maintainer can trigger CI by commenting
`/ok to test <head-sha>`; repeat this after a new push if the PR remains untrusted.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

@pzelasko @nithinraok

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -36,7 +36,7 @@ dependencies = [
     "smart-open",
     "tensorboard",
     "text-unidecode",
-    "torch>=2.6.0",
+    "torch>=2.7.0",
```
[the remark this answers is not in the document]
Comment (MahmoudAshraf97) [reply pyproject.toml:38]: all commits are signed off

File: nemo/collections/asr/parts/triton/subsampling.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+                    batch_base,
+                    mel_time_stride,
+                    relu_out_row,
+                    relu_out_col,
+                    pos_ok,
+                    this_mel_len,
+                    mel_freq,
+                    pad_start,
+                )
+                relu_out = tl.dot(conv_taps, feats) + conv_bias[:, None]
```
[the remark this answers is not in the document]
Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/triton/subsampling.py]: Most training is in either mixed precision or half precision, where TF32 has no effect at all, regardless leaving this for the maintainers call

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 3]
```diff
@@ -345,6 +348,7 @@ def __init__(
         sync_max_audio_length: bool = True,
         rope_base: float = 10000.0,
         rotary_fraction: float = 1.0,
+        use_triton: bool | None = None,
```
[the remark this answers is not in the document]
Comment (MahmoudAshraf97) [reply nemo/collections/asr/modules/conformer_encoder.py:351]: this is similar to the `use_pytorch_sdpa` option, where it only exists in the module docstring and the configs only, not the documentation

[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/modules/conformer_encoder.py:351]: yes please add to docs. we should have added `use_pytorch_sdpa` info to docs as well.

File: nemo/collections/asr/parts/submodules/subsampling.py [hunk 4]
```diff
@@ -418,6 +422,15 @@ def __init__(
...
         self.conv = MaskedConvSequential(*layers)
 
+        # The kernels implement `dw_striding`'s layout, [conv, act] + (sampling_num - 1) x
+        # [dw, pw, act], with ReLU baked in; a factor of 2 stops after [conv, act], leaving no
+        # depthwise to fuse.
+        if use_triton is None:
+            use_triton = TRITON_AVAILABLE
+        self.conv.fuse_triton = (
+            use_triton and subsampling == 'dw_striding' and self._sampling_num >= 2 and isinstance(activation, nn.ReLU)
+        )
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/subsampling.py]: but if user requests use_triton and striding, then code currently silently skips use of triton. Could you raise warning instead.

File: nemo/collections/asr/parts/submodules/subsampling.py [hunk 5]
```diff
@@ -19,6 +19,9 @@
 from torch.nn import LayerNorm
 
 from nemo.collections.asr.parts.submodules.causal_convs import CausalConv1D, CausalConv2D
+from nemo.collections.asr.parts.triton.depthwise_conv import dw_conv2d
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/subsampling.py]: wouldn;t this raise error, when triton is not available?

**Review by nithinraok — CHANGES_REQUESTED** [review]
Great work, couple of comments.

File: nemo/collections/asr/parts/submodules/subsampling.py (hunk 5 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/submodules/subsampling.py]: No, since all triton code in that file are guarded with `TRITON_AVAILABLE` flag, the alternative is to remove the gating and gate this import only

Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/submodules/subsampling.py]: rewritten the files to follow the same pattern used in the ngram triton implementation, the indentation is gone and it raises on the first forward if `use_triton=True` was explicitly requested

File: nemo/collections/asr/parts/submodules/subsampling.py (hunk 4 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/submodules/subsampling.py]: added a warning for all non-eligible configs

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MahmoudAshraf97) [reply nemo/collections/asr/modules/conformer_encoder.py:351]: added the documentation

File: nemo/collections/asr/parts/submodules/subsampling.py [hunk 6]
```diff
@@ -695,10 +712,33 @@ def calculate_conv_output_size(input_size: torch.Tensor, kernel_size: int, strid
...
+
     def forward(self, x, lengths):
         # Convert input (batch, time, features) to conv format
         x = x.unsqueeze(1)  # (batch, 1, time, features)
-        current_lengths = lengths.clone().float()
+        current_lengths = lengths
+
+        # Tracing and export cannot capture a Triton launch, the fused kernel returns no input
+        # gradient, and its weight gradients accumulate through atomics, so their summation order
+        # varies between runs.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/subsampling.py:725]: remove comment

File: pyproject.toml [hunk 7]
```diff
@@ -35,7 +35,7 @@ dependencies = [
...
```
Comment (nithinraok) [inline pyproject.toml:38]: Is 2.7 required for Triton support? Any torch version bump should go in a separate PR so we can track CI against the version change.

@chtruong814 can we upgrade this based on the container versions we currently support?

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 7 above)
Comment (MahmoudAshraf97) [reply pyproject.toml:38]: it's not required for triton in itself, but it's the minimum version that exposes `torch.compiler.is_exporting()`, which we need to skip triton during dynamo export, otherwise the export is wrong since it will skip triton kernels, given that the current default export path does not use dynamo
I can remove the torch bump from this PR and open another one, or remove it completely and accept the dynamo regression

File: nemo/collections/asr/parts/submodules/subsampling.py (hunk 6 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/submodules/subsampling.py:725]: IMO it is needed to clarify some of the skip conditions, I'll try to rewrite it in a better way as inline comments

File: pyproject.toml (hunk 7 above)
Comment (nithinraok) [reply pyproject.toml:38]: Lets do this in order:
1. Raise another PR to bump torch version
2. rebase/update this branch on top of it once previous PR merged.

Comment (MahmoudAshraf97) [reply pyproject.toml:38]: https://github.com/NVIDIA-NeMo/Speech/pull/16138

**Review by nithinraok — DISMISSED** [review]

**Review by pzelasko — APPROVED** [review]
Approving the diff changes from last commit, since Nithin approved the PR previously.
